META6.json

```diff
@@ -1,10 +1,20 @@
 {
+    "meta-version": 1,
     "perl"      : "6.*",
     "name"      : "OpenSSL",
     "version"   : "0.1.22",
     "author"    : "github:sergot",
+    "license"   : "MIT",
     "description"   : "OpenSSL bindings",
-    "depends"   : [],
+    "depends"   : {
```

Comment:
This has some issues.

The this-or-that method of declaring a dependency using array notation is not implemented. Last time I mentioned implementing this you seemed to have some reservations about it, although I don't remember why -- only that I was in favor of it (just nobody got around to implementing).

The hash form of declaration doesn't accept the string form of names. So using "ssl:from<native>:ver<1.1>" probably isn't going to work.

The format of `depends` is off. Unimportantly I would recommend using the array-of-hash format, but if not then this is still missing a level (we specced `runtime` etc). More importantly is there is no less-redundant way to properly express what you want (yet). Theoretically it could be like
```
    "depends" : [
        {
            "from" : "native",
            "name" : {
                "by-kernel.name": {
                    "win32": "ssleay32",
                    "" : "ssl"
                }
            }
        },
        {
            "from" : "native",
            "name" : {
                "by-kernel.name": {
                    "win32": "libeay32",
                    "":"ssl"
                }
            }
        }
    ]
```
but note how we currently *have* to declare a dependency -- i.e. `ssl` is listed in the second dependency block even though it is redundant against the first block.

Reply:
Oddly zef didn't seem to care about the dependencies when I tried. But maybe I just screwed up my tests. Was a bit distracted by being at FOSDEM and all.

I can't remember what my reservations were about the array notation. Actually I first thought we overlooked this use case and was pleasantly surprised when I found it it S22: "To indicate alternatives, it is possible to specify a list of use strings, instead of just a single use string". If I were asked to review the spec right now, I'd probably not like the implicitness. Usually when you see a list, it's ANDed. You wouldn't assume that your shopping list means "buy beer OR burgers". But then, why would you put a nested list into your list of dependencies? We could spell it `{"alternatives": [...]}` instead to make it more self documenting. If we could think of another combinator besides AND and OR, that'd make more sense. Or are there other reasonable search strategies besides "the first of these you can find"? If there are, that'd would be an argument for the more elaborate version.

The depends block I specified actually **is** in array-of-hash format. On e.g. Linux it will collapse to
```
"depends": [
    ["ssl:from<native>:ver<1.1>", "ssl:from<native>:ver<1.0>"],
    ["crypto:from<native>:ver<1.1>", "crypto:from<native>:ver<1.0>"]
]
```
At least if zef applies system-collapse to the depends value directly.

If you're OK with it I can implement the missing bits in zef. At this point I'm already so many layers deep into yak shaving that one more won't matter ;) We have already done all the hard work. It's about time we reap the benefits and get this stuff used.

Reply:
> The depends block I specified actually is in array-of-hash format. On e.g. Linux it will collapse to

I'm not sure this (assumption?) is correct. Collapse does not occur on the top level itself AFAIK. See [Inline::Python example](https://github.com/niner/Inline-Python/blob/53a1c2ca1751b5928909d929493e5ed29a721d02/META6.json#L10-L12) and [Zef's implementation](https://github.com/ugexe/zef/blob/7bbc6638aa62fa5535aad77252f9a3bc61e87ce1/lib/Zef/Distribution.pm6#L72-L74). If what you did actually collapses as you say then that would seem like a coincidence.

Reply:
I imagine the issue with alternative syntax was how does one then do both `AND` and `OR`

`"depends" : ["Foo1", ["Foo2", { { Foo::MaybeThis => 1 }, { Foo::MaybeThat => 1 }}]`